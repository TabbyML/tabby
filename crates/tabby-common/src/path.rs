use std::{cell::Cell, env, path::PathBuf, sync::Mutex};

use lazy_static::lazy_static;

lazy_static! {
    static ref TABBY_ROOT: Mutex<Cell<PathBuf>> = {
        Mutex::new(Cell::new(match env::var("TABBY_ROOT") {
            Ok(x) => PathBuf::from(x),
            Err(_) => home::home_dir().unwrap().join(".tabby"),
        }))
    };
    static ref TABBY_MODEL_CACHE_ROOT: Option<PathBuf> =
        env::var("TABBY_MODEL_CACHE_ROOT").ok().map(PathBuf::from);
}

#[cfg(any(feature = "testutils", test))]
pub fn set_tabby_root(path: PathBuf) {
    println!("SET TABBY ROOT: '{}'", path.display());
    let cell = TABBY_ROOT.lock().unwrap();
    cell.replace(path);
}

pub fn tabby_root() -> PathBuf {
    let mut cell = TABBY_ROOT.lock().unwrap();
    cell.get_mut().clone()
}

pub fn config_file() -> PathBuf {
    tabby_root().join("config.toml")
}

pub fn usage_id_file() -> PathBuf {
    tabby_root().join("usage_anonymous_id")
}

pub fn repositories_dir() -> PathBuf {
    tabby_root().join("repositories")
}

pub fn index_dir() -> PathBuf {
    tabby_root().join("index")
}

pub fn models_dir() -> PathBuf {
    if let Some(cache_root) = &*TABBY_MODEL_CACHE_ROOT {
        cache_root.clone()
    } else {
        tabby_root().join("models")
    }
}

pub fn events_dir() -> PathBuf {
    tabby_root().join("events")
}

mod registry {}
