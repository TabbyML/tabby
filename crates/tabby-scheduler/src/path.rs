use std::path::PathBuf;

use tabby_common::path::tabby_root;

pub fn incremental_repository_store() -> PathBuf {
    tabby_root().join("incremental-repositorries.kv")
}
