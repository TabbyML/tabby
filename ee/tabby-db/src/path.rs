use std::path::PathBuf;

use tabby_common::path::tabby_root;

fn tabby_ee_root() -> PathBuf {
    tabby_root().join("ee")
}

pub fn db_file() -> PathBuf {
    #[cfg(feature = "prod")]
    {
        tabby_ee_root().join("db.sqlite")
    }
    #[cfg(not(feature = "prod"))]
    {
        tabby_ee_root().join("dev-db.sqlite")
    }
}
