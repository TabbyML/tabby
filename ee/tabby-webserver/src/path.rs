use std::path::PathBuf;

use tabby_common::path::tabby_root;

pub fn tabby_ee_root() -> PathBuf {
    tabby_root().join("ee")
}

pub fn db_file() -> PathBuf {
    #[cfg(feature = "prod-db")]
    {
        tabby_ee_root().join("db.sqlite")
    }
    #[cfg(not(feature = "prod-db"))]
    {
        tabby_ee_root().join("dev-db.sqlite")
    }
}

pub fn job_db_file() -> PathBuf {
    tabby_ee_root().join("job.sqlite")
}
