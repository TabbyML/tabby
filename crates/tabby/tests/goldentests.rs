use std::path::PathBuf;

use lazy_static::lazy_static;
use serde_json::json;
use serial_test::serial;
use tokio::{
    process::Command,
    time::{sleep, Duration},
};

lazy_static! {
    static ref CLIENT: reqwest::Client = reqwest::Client::new();
}

fn workspace_dir() -> PathBuf {
    let output = std::process::Command::new(env!("CARGO"))
        .arg("locate-project")
        .arg("--workspace")
        .arg("--message-format=plain")
        .output()
        .unwrap()
        .stdout;
    let cargo_path = std::path::Path::new(std::str::from_utf8(&output).expect("Valid path").trim());
    cargo_path
        .parent()
        .expect("Path must have a parent folder")
        .to_path_buf()
}

fn tabby_path() -> PathBuf {
    workspace_dir().join("target/debug/tabby")
}

fn initialize_server(gpu_device: Option<&str>) {
    let mut cmd = Command::new(tabby_path());
    cmd.arg("serve")
        .arg("--model")
        .arg("TabbyML/StarCoder-1B")
        .arg("--no-webserver")
        .arg("--port")
        .arg("9090")
        .kill_on_drop(true);

    if let Some(gpu_device) = gpu_device {
        cmd.arg("--device").arg(gpu_device);
    }

    tokio::task::spawn(async move {
        cmd.spawn()
            .expect("Failed to start server")
            .wait()
            .await
            .expect("Failed to start server");
    });
}

async fn wait_for_server(device: Option<&str>) {
    initialize_server(device);

    loop {
        println!("Waiting for server to start...");
        let is_ok = reqwest::get("http://127.0.0.1:9090/v1/health")
            .await
            .is_ok();
        if is_ok {
            break;
        } else {
            sleep(Duration::from_secs(5)).await;
        }
    }
}

async fn golden_test(body: serde_json::Value) -> serde_json::Value {
    let mut body = body.clone();
    body.as_object_mut().unwrap().insert(
        "debug_options".to_owned(),
        json!({
            "disable_retrieval_augmented_code_completion": true
        }),
    );

    let actual: serde_json::Value = CLIENT
        .post("http://127.0.0.1:9090/v1/completions")
        .json(&body)
        .send()
        .await
        .unwrap()
        .json()
        .await
        .unwrap();
    actual
}

macro_rules! assert_golden {
    ($expr:expr) => {
        insta::assert_yaml_snapshot!(golden_test($expr).await, {
            ".id" => "test-id"
        });
    }
}

#[cfg(all(target_os = "macos", target_arch = "aarch64"))]
#[tokio::test]
#[serial]
async fn run_golden_tests() {
    wait_for_server(Some("metal")).await;

    assert_golden!(json!({
            "language": "python",
            "seed": 0,
            "segments": {
                "prefix": "def fib(n):\n    ",
                "suffix": "\n        return fib(n - 1) + fib(n - 2)"
            }
    }));

    assert_golden!(json!({
            "language": "python",
            "seed": 0,
            "segments": {
                "prefix": "import datetime\n\ndef parse_expenses(expenses_string):\n    \"\"\"Parse the list of expenses and return the list of triples (date, value, currency).\n    Ignore lines starting with #.\n    Parse the date using datetime.\n    Example expenses_string:\n        2016-01-02 -34.01 USD\n        2016-01-03 2.59 DKK\n        2016-01-03 -2.72 EUR\n    \"\"\"\n    for line in expenses_string.split('\\n'):\n        "
            }
    }));
}

#[tokio::test]
#[serial]
async fn run_golden_tests_cpu() {
    wait_for_server(Some("cpu")).await;

    assert_golden!(json!({
            "language": "python",
            "seed": 0,
            "segments": {
                "prefix": "def is_prime(n):\n",
            }
    }));

    assert_golden!(json!({
            "language": "python",
            "seed": 0,
            "segments": {
                "prefix": "def char_frequencies(str):\n  freqs = {}\n  ",
            }
    }));
}
