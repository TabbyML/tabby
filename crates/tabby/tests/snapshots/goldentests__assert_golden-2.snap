---
source: crates/tabby/tests/goldentests.rs
expression: golden_test(body).await
---
id: test-id
choices:
  - index: 0
    text: "if line.startswith('#'):\n            continue\n        date, value, currency = line.split()\n        date = datetime.datetime.strptime(date, '%Y-%m-%d')\n        yield date, float(value), currency"
debug_data: {}

